src/lib.rs

```diff
@@ -245,6 +245,18 @@ impl Peripherals {
         Some(Peripherals::new(p))
     }
 
+    #[cfg(feature = "rtic")]
+    /// Steal all of the HAL's peripherals.
+    ///
+    /// # Safety
+    ///
+    /// NOTE: This constructor is only intended for use with the `rtic` crate. This is **not** an
+    /// alternative to the `take` constructor. The `take` constructor sets the system timer
+    /// interrupt while this constructor does not seeing as `rtic` will take care of this for us.
+    pub unsafe fn steal() -> Self {
+        Self::new(hal::Peripherals::steal())
+    }
```

Comment:
> ...`rtic` will take care of this for us.

I wasn't 100% sure about this - is this the case? Currently reading through the `rtic` book but not far enough through to be sure.

Reply:
Not too familiar with RTIC either, but I think this is the right call. In fact, we may need to go further!

From the [Timer queue](https://rtic.rs/0.5/book/en/by-example/timer-queue.html) section,

> When the schedule API is being used the runtime internally uses the SysTick interrupt handler and the system timer peripheral (SYST) so neither can be used by the application.

The `teensy4-bsp` implements `SysTick`, too.

https://github.com/mciantyre/teensy4-rs/blob/1d971e4e99d2bece40f362d07da7fcc85363217e/src/lib.rs#L348-L353

If we use RTCI's schedule API, this will result in a duplicate definition error. We can reproduce this with a few modifications to the example; see f612922 for more information.

I think the simplest way to work around this is to *not* define `SysTick` in the BSP if we're using the `"rtic"` feature. What do you think?

Reply:
Ahhh soon after I posted I also tried extending the example with a couple more lines that caused the compiled `.hex` files to be unreadable by `teensy-cli-loader` - perhaps this is related? Here's a minimal example that demonstrates the issue:

```rust
#![no_std]
#![no_main]

use embedded_hal::digital::v2::{OutputPin, ToggleableOutputPin};
use panic_halt as _;
use teensy4_bsp as bsp;

#[rtic::app(device = teensy4_bsp, peripherals = true)]
const APP: () = {
    struct Resources {
        led: bsp::LED,
    }

    #[init]
    fn init(mut cx: init::Context) -> init::LateResources {
        // Prepare the LED.
        let mut led = bsp::configure_led(&mut cx.device.gpr, cx.device.pins.p13);
        led.set_high().unwrap();
        init::LateResources { led }
    }

    #[idle(resources = [led])]
    fn idle(cx: idle::Context) -> ! {
        cx.resources.led.toggle().unwrap(); // Removing this line, everything works fine.
        loop {
            core::sync::atomic::spin_loop_hint();
        }
    }
};
```

The only extra step going on here is attempting to store the `LED` in `Resources` following [this part](https://rtic.rs/0.5/book/en/by-example/resources.html#late-resources) of the guide.

I can compile the above and use `arm-none-eabi-objcopy` to create the `.hex`, however when attempting to load it onto the teensy the loader produces the following output:

```
Teensy Loader, Command Line, Version 2.1
Warning, HEX parse error line 2
error reading intel hex file "out/morph_fw_module.hex"
```

Here's [the broken .hex file](https://drive.google.com/file/d/17CkoUoiowEZ8yYf-QVuU59A_WcVXOhhW/view?usp=sharing) in case you're curious to take a look.

RTIC also appears to add the generated code to `target/` too - here's the output after running through rustfmt:

<details>
  <summary>RTIC expansion</summary>

  ```rust
  #[allow(non_snake_case)]
  fn init(mut cx: init::Context) -> init::LateResources {
      let mut led = bsp::configure_led(&mut cx.device.gpr, cx.device.pins.p13);
      led.set_high().unwrap();
      init::LateResources { led }
  }
  #[allow(non_snake_case)]
  fn idle(cx: idle::Context) -> ! {
      use rtic::Mutex as _;
      cx.resources.led.toggle().unwrap();
      loop {
          core::sync::atomic::spin_loop_hint();
      }
  }
  #[doc = r" Resources initialized at runtime"]
  #[allow(non_snake_case)]
  pub struct initLateResources {
      pub led: bsp::LED,
  }
  #[allow(non_snake_case)]
  #[doc = "Initialization function"]
  pub mod init {
      #[doc(inline)]
      pub use super::initLateResources as LateResources;
      #[doc = r" Execution context"]
      pub struct Context {
          #[doc = r" Core (Cortex-M) peripherals"]
          pub core: rtic::export::Peripherals,
          #[doc = r" Device peripherals"]
          pub device: teensy4_bsp::Peripherals,
      }
      impl Context {
          #[inline(always)]
          pub unsafe fn new(core: rtic::export::Peripherals) -> Self {
              Context {
                  device: teensy4_bsp::Peripherals::steal(),
                  core,
              }
          }
      }
  }
  #[allow(non_snake_case)]
  #[doc = "Resources `idle` has access to"]
  pub struct idleResources {
      pub led: &'static mut bsp::LED,
  }
  #[allow(non_snake_case)]
  #[doc = "Idle loop"]
  pub mod idle {
      #[doc(inline)]
      pub use super::idleResources as Resources;
      #[doc = r" Execution context"]
      pub struct Context {
          #[doc = r" Resources this task has access to"]
          pub resources: Resources,
      }
      impl Context {
          #[inline(always)]
          pub unsafe fn new(priority: &rtic::export::Priority) -> Self {
              Context {
                  resources: Resources::new(priority),
              }
          }
      }
  }
  #[doc = r" Implementation details"]
  const APP: () = {
      #[doc = r" Always include the device crate which contains the vector table"]
      use teensy4_bsp as _;
      #[cfg(core = "1")]
      compile_error!("specified 1 core but tried to compile for more than 1 core");
      impl idleResources {
          #[inline(always)]
          unsafe fn new(priority: &rtic::export::Priority) -> Self {
              idleResources {
                  led: &mut *led.as_mut_ptr(),
              }
          }
      }
      #[allow(non_upper_case_globals)]
      #[link_section = ".uninit.rtic0"]
      static mut led: core::mem::MaybeUninit<bsp::LED> = core::mem::MaybeUninit::uninit();
      #[no_mangle]
      unsafe extern "C" fn main() -> ! {
          let _TODO: () = ();
          rtic::export::interrupt::disable();
          let mut core: rtic::export::Peripherals = core::mem::transmute(());
          let late = crate::init(init::Context::new(core.into()));
          led.as_mut_ptr().write(late.led);
          rtic::export::interrupt::enable();
          crate::idle(idle::Context::new(&rtic::export::Priority::new(0)))
      }
  };
  ```
</details>

I'll have a go at removing the duplicate `SysTick` definition to see if it fixes things and report back.

> I think the simplest way to work around this is to not define SysTick in the BSP if we're using the "rtic" feature. 

Yeah this sounds like the easiest approach. Folks tend to recommend that features remain "additive" (rust-lang/cargo#4328), but there is precedent for them being useful to select between a set of backends and I think this use-case is along those same lines.

Reply:
OK, so I tried my updated LED example above after removing the teensy4-bsp `SysTick` definition from `teensy4_bsp` when the `rtic` feature is enabled but no luck - I still get the same `Warning, HEX parse error line 2` error as above.

I also tried running your example at https://github.com/mciantyre/teensy4-rs/commit/f612922ff2b729d18b78ae0af3707a0b9da7621b and did not run into any linker errors, but did run into a couple of new compiler errors:

```
   Compiling teensy4-bsp v0.1.0 (/home/mindtree/programming/rust/teensy4-rs)
error[E0433]: failed to resolve: could not find `Interrupt` in `teensy4_bsp`
  --> examples/rtic_led2.rs:19:1
   |
19 | #[rtic::app(device = teensy4_bsp, peripherals = true, monotonic = rtic::cyccnt::CYCCNT)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ could not find `Interrupt` in `teensy4_bsp`
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find value `NVIC_PRIO_BITS` in crate `teensy4_bsp`
  --> examples/rtic_led2.rs:19:1
   |
19 | #[rtic::app(device = teensy4_bsp, peripherals = true, monotonic = rtic::cyccnt::CYCCNT)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `teensy4_bsp`
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors
```

Adding the following to the `teensy4_bsp` root seems to solve these.

```rust
// `rtic` expects these in the root.
#[cfg(feature = "rtic")]
pub use hal::ral::{interrupt as Interrupt, NVIC_PRIO_BITS};
```

After this, it appears I can upload and run your program successfully!

Perhaps the next step is to try and get a basic timer interrupt working, however first we may need to solve the issue I'm running into (where using the `led` outside of `init` seems to result in an invalid `.hex`) in order to be able to tell if our timer is working or not :thinking:

Reply:
> I also tried extending the example with a couple more lines that caused the compiled .hex files to be unreadable by teensy-cli-loader

We'll need to specify an unloaded `.uninit` section in the `teensy4-rt`'s linker script:

```
    .uninit (NOLOAD) :
    {
        *(.uninit .uninit.*);
        . = ALIGN(16);
    } > DTCM
```

I'm sorry for the trouble! We found this when we were [first exploring RTFM](https://github.com/mciantyre/teensy4-rs/pull/52#issuecomment-599255134), but I never followed-up on it. I've committed this directly to `master` (ca100f3), so it doesn't happen again.

Your late-resources example should work with the new link section; I tested in 31b140c, and the LED turns on (commenting-out the LED modifications in `init()`).

> I also tried running your example at f612922 [...] but did run into a couple of new compiler errors

Again sorry 😛 I should have noted that f612922 was the child of commit 8e1fdce, which exposed those two symbols.

Reply:
> Folks tend to recommend that features remain "additive" [...] 

Now might be a good time to make that true. I've been thinking about guarding the USB logging feature behind a feature-flag, so that users who don't want the USB stack don't need it. To keep things backwards-compatible, the `teensy4-bsp`'s default feature will use `"usb-logging"`. `"usb-logging"` depends on `"systick"`, which is another default feature. Effectively, it would look like.

```toml
[features]
default = ["usb-logging", "systick"]
usb-logging = ["systick"]
systick = []
rtic = []
```

With this approach, we would be able to support RTIC with the `"rtic"` feature flag *after disabling all default features*:

```
[dependencies]
cortex-m-rtic = "0.5"

[dependencies.teensy4-bsp]
default-features = false
features = ["rtic"]
```

How would that sound? Everything is still "additive," we still have all the support we want, and today's users don't notice a thing.

Reply:
> Your late-resources example should work with the new link section; I tested in 31b140c, and the LED turns on (commenting-out the LED modifications in init()).

Ahhh great to hear its a known issue! Much better than an unknown one :) Could you share any insight into how you tracked this down? I imagine it would have taken me a while to come across this solution continuing hacking at this on my own, though perhaps this is just my severe lack of experience working with teensy and linker scripts showing! Thanks btw, I've rebased this PR onto master again - I'll test my example again locally too once I'm at my teensy again.

> I've been thinking about guarding the USB logging feature behind a feature-flag, so that users who don't want the USB stack don't need it.

Awesome! I'm hoping to use the teensy USB serial interface for interfacing with some control software, so this was going to be one of my next issues. The proposed feature set sounds like a great solution!

I noticed there's a `USB-OTG2` in the `imxrt-ral` - is there a second USB interface I could possibly expose somehow for debugging?

While we're on the topic of teensy 4.0 USB serial - do you happen to know what kind of transfer speeds can be achieved? I've noticed that the baud rate is normally specified as `9600`, however the comment in [the arduino example](https://www.pjrc.com/teensy/td_serial.html) mentions this value is ignored and that the speed is always 12 Mbit/s, while the teensy 4 advertises "high speed" USB (480 Mbit/sec). I'm anticipating some potentially high rates in my project that may end up exceeding 12 Mbit/s so I'm curious about whether or not the "high-speed" is even achievable via the USB serial or if I'm misinterpreting the 480 Mbit/sec and its accessibility.

Reply:
> Could you share any insight into how you tracked this down?

<details>
<summary>How I stumbled upon a fix...</summary>

The Teensy loader indicates an error when it doesn't like a HEX file. We can study the loader's source code to understand what kind of records the loader accepts, and what it rejects. But, that won't help us understand what's wrong with our specific program. Intel HEX files aren't particularly fun to explore. ELF files, which become our HEX files, do have helpful information. Let's summarize the troublesome program's sections...

```
cargo objdump --example rtic_led --no-default-features --features=rtic -- -h

rtic_led:	file format ELF32-arm-little

Sections:
Idx Name          Size     VMA              LMA              Type
  0               00000000 0000000000000000 0000000060001780
  1 .boot         00001780 0000000060000000 0000000060000000 TEXT
  2 .text         00002ad0 0000000000000000 0000000060001780 TEXT
  3 .padding      00005530 0000000000002ad0 0000000000002ad0 TEXT BSS
  4 .data         00000200 0000000020000000 0000000060004250 DATA
  5 .uninit.rtic0 00000004 0000000000008000 000000004000c250 DATA
  6 .bss          00001300 0000000020001000 0000000060005250 BSS
  7 .dma          00003060 0000000020200000 0000000020200000 BSS
```

Everything looks reasonable for our processor's memory map (described in the iMXRT1062 reference manual). But, we notice a new section, `.uninit.rtic0`. Kudos to the RTIC team for putting 'rtic' in the name, so we know where it comes from. For some reason, it's placed at `0x4000c250` which is peripheral memory.

We should tell the linker where to place the `.uninit` section. Once we do, our memory will look like this:

```
cargo objdump --example rtic_led --no-default-features --features=rtic -- -h

Sections:
Idx Name          Size     VMA              LMA              Type
  0               00000000 0000000000000000 0000000060001780
  1 .boot         00001780 0000000060000000 0000000060000000 TEXT
  2 .text         00002ad0 0000000000000000 0000000060001780 TEXT
  3 .padding      00005530 0000000000002ad0 0000000000002ad0 TEXT BSS
  4 .data         00000200 0000000020000000 0000000060004250 DATA
  5 .bss          00001300 0000000020001000 0000000060005250 BSS
  6 .dma          00003060 0000000020200000 0000000020200000 BSS
  7 .uninit       00000010 0000000020002300 0000000020002300 BSS
```

Now, `.uninit` is properly in the DTCM memory region. It's uninitialized, and RTIC will initialize the variables that end up there.

Since the section isn't loaded, it won't appear in the output HEX file. No troublesome records, and the loader will happily download our application.
</details>

> I noticed there's a USB-OTG2 in the imxrt-ral - is there a second USB interface I could possibly expose somehow for debugging?

<details>
<summary>There might be...</summary>

[The product page](https://www.pjrc.com/store/teensy40.html) shows the pinout and pin capabilities for the Teensy 4.0. If we look on the underside of the Teensy 4.0's USB connector, there are D+ and D- pads. That might be for the USB2 peripheral. I've not explored it. Could be worth studying Paul's [USBHost_t36](https://github.com/PaulStoffregen/USBHost_t36) library.
</details>

> do you happen to know what kind of transfer speeds can be achieved

<details>
<summary>This is a high-speed USB stack</summary>
The C driver is written to support both full- and high-speeds. My computer identifies the Teensy it as a USB 2 device. Can't say for sure, but that document may be out of date.
</details>

Reply:
> How I stumbled upon a fix...

Amazing. Thanks so much for the detailed insight!!

> My computer identifies the Teensy it as a USB 2 device. Can't say for sure, but that document may be out of date.

Thanks for the tips. I think my next foray will be attempting to get USB working alongside `rtic`. I think first I'd like to try and get your logging via USB serial working (especially as it would make all following work easier to debug)! Then I'd like to test what kind of throughput I can get when moving large amounts of data between my teensy and laptop and measure the round-trip.